### PR TITLE
BF: for degenerate polygons, add CLOSEPOLY vertex

### LINKED ISSUE #17975: Computing the bounding box of a degenerate polygon throws an error
With 3.3.0, computing the bounding box of a degenerate (one-point) polygon throws an error

**Code for reproduction**

```python
from matplotlib.patches import Polygon                                                                                                            
pp=Polygon([[0.,0.]])                                                                                                                             
pp.get_extents()                                                                                                                                  
```

**Actual outcome**

```
ValueError                                Traceback (most recent call last)
<ipython-input-3-1312a177eac2> in <module>
----> 1 pp.get_extents()

/usr/lib/python3.8/site-packages/matplotlib/patches.py in get_extents(self)
    254         Return the `Patch`'s axis-aligned extents as a `~.transforms.Bbox`.
    255         """
--> 256         return self.get_path().get_extents(self.get_transform())
    257 
    258     def get_transform(self):

/usr/lib/python3.8/site-packages/matplotlib/path.py in get_extents(self, transform, **kwargs)
    590             self = transform.transform_path(self)
    591         bbox = Bbox.null()
--> 592         for curve, code in self.iter_bezier(**kwargs):
    593             # places where the derivative is zero can be extrema
    594             _, dzeros = curve.axis_aligned_extrema()

/usr/lib/python3.8/site-packages/matplotlib/path.py in iter_bezier(self, **kwargs)
    442             if first_vert is None:
    443                 if code != Path.MOVETO:
--> 444                     raise ValueError("Malformed path, must start with MOVETO.")
    445             if code == Path.MOVETO:  # a point is like "CURVE1"
    446                 first_vert = verts

ValueError: Malformed path, must start with MOVETO.
```

**Expected outcome**

With 3.2:
```
Bbox([[inf, inf], [-inf, -inf]])
```

**Matplotlib version**
  * Operating system: Linux x86_64
  * Matplotlib version: 3.3.0
  * Matplotlib backend (`print(matplotlib.get_backend())`): Qt5Agg
  * Python version: 3.8.4

### PR DESCRIPTION
**Pull request by brunobeltran**
## PR Summary

Fixes #17975. New `iter_bezier` code exposed an existing bug in `patches.Polygon.set_xy`.

The check `xy[0] != xy[-1]` was meant to check if the user had already included an extra `CLOSEPOLY` vertex for us in the list of vertices (`shape(xy) = (N, 2)`). *But* for single-vertex (degenerate) `Polygon`s, it produced a false positive (since `xy[0] == xy[-1]` by definition), leading to malformed `Path`s that look like:

```python
In [1]: from matplotlib.patches import Polygon
In [2]: pp = Polygon([[0, 0]])
In [3]: pp.get_path()
Out[3]: Path(array([[0., 0.]]), array([79], dtype=uint8))
```

## PR Checklist

- [ ] Has Pytest style unit tests
- [x] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/next_api_changes/* if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/patches.py [hunk 1]
```diff
@@ -1066,10 +1066,18 @@ def set_xy(self, xy):
...
+        ``CLOSEPOLY`` vertex, and add it ourselves.
         """
         xy = np.asarray(xy)
+        N, _ = xy.shape
         if self._closed:
-            if len(xy) and (xy[0] != xy[-1]).any():
+            if N == 1 or N > 1 and (xy[0] != xy[-1]).any():
                 xy = np.concatenate([xy, [xy[0]]])
         else:
             if len(xy) > 2 and (xy[0] == xy[-1]).all():
```
Comment (QuLogic) [inline lib/matplotlib/patches.py:1083]: Since you have `N` now, I expect you want to use it here too?

Comment (brunobeltran) [reply lib/matplotlib/patches.py]: Clearly I need to get more sleep, fixed.

**Comment by brunobeltran** [comment]
> Add a test?

Couldn't decide what to test initially. Settled on testing that for a single point (x, y), the extents of a degenerate polygon with only that point as its vertices has the extents `Bbox([[x, y], [x, y]])`, which is the behavior where we actually differ from the (buggy) 3.2 behavior.

**Comment by jklymak** [comment]
I'm sure this was covered in the other PR, but what happens if x=np.Nan?

**Comment by brunobeltran** [comment]
tl;dr: NaN works.

The `Polygon` class doesn't complain about NaN's or clean them explicitly (currently. You might argue that it should).

So `Polygon([[np.nan, 0]])` creates a path that *would* be correct if there were no NaN's there: `Path([[np.nan, 0], [np.nan, 0]], [MOVETO, CLOSEPOLY])`. The new `Path.get_extents` blindly adds all possible extremal values to a `Bbox.null()` (which in this case are all NaN). Adding a NaN to a Bbox is a no-op, (the call chain ends in `_path.update_path_extents`, which works on a `PathNaNRemover<agg:conv_transform<PathIterator>>`), so the `Bbox` stays null, correctly indicating there are no non-NaN points in the polygon.

**Comment by brunobeltran** [comment]
CI fail is due to "error allocating resources" on our Mac box?

**Comment by jklymak** [comment]
... the mac CI on azure seems to have issues

File: lib/matplotlib/patches.py [hunk 2]
```diff
@@ -1066,13 +1066,21 @@ def set_xy(self, xy):
...
+        Unlike `~.path.Path`, we do not ignore the last input vertex. If the
+        polygon is meant to be closed, and the last point of the polygon is not
+        equal to the first, we assume that the user has not explicitly passed a
+        ``CLOSEPOLY`` vertex, and add it ourselves.
         """
         xy = np.asarray(xy)
+        N, _ = xy.shape
         if self._closed:
-            if len(xy) and (xy[0] != xy[-1]).any():
+            if N == 1 or N > 1 and (xy[0] != xy[-1]).any():
```
Comment (dstansby) [inline lib/matplotlib/patches.py:1080]: Could you add a quick comment about the `if` statements you've changed just to say what they're doing? And maybe `nverts` is a better variable name than just `N`?

Comment (brunobeltran) [reply lib/matplotlib/patches.py]: I was reticent to add these comments, because technically they are very brittle checks as-written, since we don't guarantee anything about the coordinates of a CLOSEPOLY vertex, but this code uses the assumption that these vertices will "correctly" match the start point of the curve.

In particular, both Python (e.g. `Path.iter_bezier`) and C++ code (i.e. handlers for `agg:path_flags_close`) actively ignore this vertex's value, because it's known to hold "trash" sometimes, especially after cleaning NaN arrays, etc (weird stuff like #17885 and #17914).

But, I added comments that make it more clear what the if statements are *nominally* doing, since you think it's a good idea. I personally like `N` since it matches what all the `Polygon` docstrings use, but I agree that `nverts` is easier to read if you're skimming the code only.
